kong/llm/init.lua

```diff
@@ -252,6 +381,159 @@ function _M.is_compatible(request, route_type)
   return false, fmt("[%s] message format is not compatible with [%s] route type", format, route_type)
 end
 
+function _M:handle_streaming_request(body)
+  -- convert it to the specified driver format
+  local request, _, err = self.driver.to_format(body, self.conf.model, self.conf.route_type)
+  if err then
+    return internal_server_error(err)
+  end
+
+  -- run the shared logging/analytics/auth function
+  ai_shared.pre_request(self.conf, request)
+
+  local prompt_tokens = 0
+  local err
+  if not ai_shared.streaming_has_token_counts[self.conf.model.provider] then
+    -- Estimate the cost using KONG CX's counter implementation
+    prompt_tokens, err = self:calculate_cost(request, {}, 1.8, 0)
+    if err then
+      return internal_server_error("unable to estimate request token cost: " .. err)
+    end
+  end
+
+  -- send it to the ai service
+  local res, _, err, httpc = self.driver.subrequest(request, self.conf, self.http_opts, true)
+  if err then
+    return internal_server_error("failed to connect to " .. self.conf.model.provider .. " for streaming: " .. err)
+  end
+  if res.status ~= 200 then
+    err = "bad status code whilst opening streaming to " .. self.conf.model.provider .. ": " .. res.status
+    ngx.log(ngx.WARN, err)
+    return bad_request(err)
+  end
+
+  -- get a big enough buffered ready to make sure we rip the entire chunk(s) each time
+  local reader = res.body_reader
+  local buffer_size = 35536
+  local events
+
+  -- we create a fake "kong response" table to pass to the telemetry handler later
+  local telemetry_fake_table = {
+    response = buf:new(),
+    usage = {
+      prompt_tokens = prompt_tokens,
+      completion_tokens = 0,
+      total_tokens = 0,
+    },
+  }
+
+  ngx.status = 200
+  ngx.header["Content-Type"] = "text/event-stream"
+  ngx.header["Via"] = meta._SERVER_TOKENS
+
+  for k, v in pairs(res.headers) do
+    if not streaming_skip_headers[lower(k)] then
+      ngx.header[k] = v
+    end
+  end
+
+  -- server-sent events should ALWAYS be chunk encoded.
+  -- if they aren't then... we just won't support them.
+  repeat
+    -- receive next chunk
+    local buffer, err = reader(buffer_size)
+    if err then
+        ngx.log(ngx.ERR, err)
+        break
+    end
+
+    if buffer then
+      -- we need to rip each message from this chunk
+      events = {}
+      for s in buffer:gmatch("[^\r\n]+") do
+        table.insert(events, s)
+      end
+
+      local metadata
+
+      -- then parse each into the standard inference format
+      for i, event in ipairs(events) do
+        local event_t
+        local token_t
+
+        -- some LLMs do a final reply with token counts, and such
+        -- so we will stash them if supported
+        local formatted, err, this_metadata = self.driver.from_format(event, self.conf.model, "stream/" .. self.conf.route_type)
```

Comment:
Performance: pre-compute this string instead of re-generating it in a loop

```lua
local route_type = "stream/" .. self.conf.route_type

-- ...

repeat
  -- ...

  for i, event in ipairs(events) do
    -- ...

    local formatted, err, this_metadata = self.driver.from_format(event, self.conf.model, route_type)
```

Reply:
Fixed in 01d8a4a1c7ee5a975832d14299bfbd228b9d4d99

Reply:
Fixed in 01d8a4a1c7ee5a975832d14299bfbd228b9d4d99